proposals/NNNN-method-keypaths.md

```diff
@@ -0,0 +1,376 @@
+# Allow Key Paths to Refer to Unapplied Instance Methods
+
+* Proposal: [SE-NNNN](NNNN-method-keypaths.md)
+* Authors: [Paul Cantrell](https://github.com/pcantrell), [Jeremy Saklad](https://github.com/Saklad5), [Filip Sakel](https://github.com/filip-sakel)
+* Review Manager: TBD
+* Status: **Awaiting implementation**
+
+## Introduction
+
+The [original key path proposal](https://github.com/apple/swift-evolution/blob/master/proposals/0161-key-paths.md)
+expressly limited key paths to be able to reference only properties and subscripts:
+
+```swift
+\Person.name                  // KeyPath<Person, String>
+\Person.pets[0]               // KeyPath<Person, Pet>
+```
+
+This proposal adds the ability for key paths to reference instance methods, optionally specifying argument names:
+
+```swift
+\Person.sing                  // KeyPath<Person, () -> Sound>
+\Person.sing(melody:lyrics:)  // KeyPath<Person, (Melody, String) -> Sound>
+```
+
+Note that these key paths do not provide argument values; they reference _unapplied_ methods, and the value they give is
+a function, not the the value that results from calling the method.
+
+Adding this capability not only removes an inconsistency in Swift, but solves pratical problems involving map/filter
```

Comment:
I think there's a grammatical error in:
`Adding this capability not only removes... but solves`

Correction:
`Adding this capability does not only remove ... but it also solves`

Reply:
“…not only removes … but solves …” is grammatically correct. Adding “but also solves” does make it flow better, though, so I’ll add that piece.